### PR TITLE
Add Tab Hierarchy

### PR DESCRIPTION
**Pull request by KaiTheRedNinja**
# Description

The idea of the tab hierarchy is to allow the user to "save" tabs, including their line number (not implemented). This would, for example, make traversing large OOP trees a lot easier since the user can essentially note down what line of a file the user deems as essential.

The UXUI can be improved. Currently, it requires a double click to open a saved tab, which differs from the Project Navigator and may not be very user friendly. Line number is also uncomfortably close to the file name.

# Related Issue

none

# Checklist

- [x] I read and understood the [contributing guide](https://github.com/AuroraEditor/AuroraEditor/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/AuroraEditor/AuroraEditor/blob/main/CODE_OF_CONDUCT.md)
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] I documented my code
- [x] Review requested

# Screenshots

Adding, nesting, and opening saved tabs
https://user-images.githubusercontent.com/88234730/190424635-8178cc5d-d5cb-40aa-81fa-26a2479364d0.mov

### CODE REVIEW

File: AuroraEditor/Base/Documents/Model/WorkspaceDocument/WorkspaceDocument+Selection.swift [hunk 1]
```diff
@@ -66,9 +77,11 @@ struct WorkspaceSelectionState: Codable {
     func getItemByTab(id: TabBarItemID) -> TabBarItemRepresentable? {
         switch id {
         case .codeEditor:
-            return self.openFileItems.first { item in
+            let path = id.id.replacingOccurrences(of: "codeEditor_", with: "")
+            // get it from the open file items, else fallback to the whole index
+            return self.openFileItems.first(where: { item in
                 item.tabID == id
-            }
+            }) ?? (try? workspace?.fileSystemClient?.getFileItem(path))
```
Comment (wdg) [inline AuroraEditor/Base/Documents/Model/WorkspaceDocument/WorkspaceDocument+Selection.swift:84]: Please try to split this, or use a guard let, this is not very readable.

File: AuroraEditor/Base/NavigatorSidebar/Model/HierarchyNavigator/TabHierarchyMenu.swift [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            self.workspace?.openTab(item: itemTab)
+        }
+    }
+
+    @objc
+    func deleteItem() {
+        guard let item = item else { return }
+        // Remove the item from its old location
+        if let recievedParentID = item.parentItem?.id {
+            for tab in workspace?.selectionState.flattenedSavedTabs ?? [] where tab.id == recievedParentID {
```
Comment (wdg) [inline AuroraEditor/Base/NavigatorSidebar/Model/HierarchyNavigator/TabHierarchyMenu.swift:78]:
```suggestion
        guard let item = item,
              let workspace = workspace else { return }
        // Remove the item from its old location
        if let recievedParentID = item.parentItem?.id {
            for tab in workspace.selectionState.flattenedSavedTabs where tab.id == recievedParentID {
```

**Comment by wdg** [comment]
Restarting the test job.

CompileSwiftSources normal x86_64 com.apple.xcode.tools.swift.compiler (in target 'AuroraEditor' from project 'AuroraEditor') [5984](https://github.com/AuroraEditor/AuroraEditor/actions/runs/3064587087/jobs/4947822965#step:7:5985) (52 failures)

**Comment by wdg** [comment]
It looks like this code does not compile in a lower version of Xcode/MacOS

```
2022-09-16T12:04:30.2619830Z /Users/runner/work/AuroraEditor/AuroraEditor/AuroraEditor/Base/TabBar/Model/TabBarItemStorage.swift:70:1: error: constrained extension must be declared on the unspecialized generic type 'Array' with constraints specified by a 'where' clause
2022-09-16T12:04:30.2620330Z extension Array<TabBarItemStorage> {
2022-09-16T12:04:30.2620560Z ^         ~~~~~~~~~~~~~~~~~~~~~~~~
2022-09-16T12:04:30.2620790Z Command EmitSwiftModule failed with a nonzero exit code
2022-09-16T12:04:30.2620960Z 
2022-09-16T12:04:30.2621550Z CompileSwift normal x86_64 
```

**Review by wdg — APPROVED** [review]

**Review by nanashili — APPROVED** [review]

**Pull request merged**